.circleci/config.yml

```diff
@@ -84,30 +93,398 @@ jobs:
+                    "text": ":npm:  *The following packages have been published to the \`$DIST_TAG\` tag*\n\n*Packages:*\n\n$MARKDOWN_VERSIONS"
+                  }
+                }
+              ]
+            }
+            EOM
+      - run:
+          name: Post to Slack on Failure
+          command: |
+            if [ -z "$SLACK_WEBHOOK_URL" ]; then
+              echo 'Cannot post failure to Slack because $SLACK_WEBHOOK_URL is not set.'
+              exit 1
+            fi
+
+            cat \<<EOM | curl -X POST -H 'Content-type: application/json' --data @- $SLACK_WEBHOOK_URL
+            {
+              "text": "A failure occurred during publishing.  For more information, see $CIRCLE_BUILD_URL.",
+              "blocks": [
+                {
+                  "type": "section",
+                  "text": {
+                    "type": "mrkdwn",
+                    "text": ":boom: *Failure during publishing!* :boom:\n\n<$CIRCLE_BUILD_URL|See the build details for more information!>"
+                  }
+                }
+              ]
+            }
+            EOM
+          when: on_fail
+
 # XXX We used to use this filter to only run a "Docs" job on docs branches.
 #     Now we use it to disable all jobs. It's unclear if there's a simpler way
 #     to do this!
-ignore_doc_branches: &ignore_doc_branches
+common_non_publish_filters: &common_non_publish_filters
   filters:
     branches:
       # If 'docs' is found, with word boundaries on either side, skip.
       ignore: /.*?\bdocs\b.*/
+    # Ensure every job has `tags` filters since the publish steps have tags.
+    # This is some wild configuration thing that's pretty hard to figure out.
+    tags:
+      only: /.*/
+
+common_publish_filters: &common_publish_filters
+  filters:
+    # Only run pre-publish and publish steps on specific tags.
+    tags:
+      only: /^publish\/[0-9]+$/
+    # We want the publish to trigger on the above tag, not any branch.
+    branches:
+      ignore: /.*/
 
 workflows:
   version: 2
-  Build and Test:
+  Build:
     jobs:
       - NodeJS 6:
-          <<: *ignore_doc_branches
+          <<: *common_non_publish_filters
       - NodeJS 8:
-          <<: *ignore_doc_branches
+          <<: *common_non_publish_filters
       - NodeJS 10:
-          <<: *ignore_doc_branches
+          <<: *common_non_publish_filters
       - NodeJS 12:
-          <<: *ignore_doc_branches
+          <<: *common_non_publish_filters
+      - Package tarballs:
+          <<: *common_non_publish_filters
+          requires:
+            - NodeJS 6
+            - NodeJS 8
+            - NodeJS 10
+            - NodeJS 12
+      - Pre-publish:
+          <<: *common_publish_filters
+          requires:
+            - NodeJS 6
+            - NodeJS 8
+            - NodeJS 10
+            - NodeJS 12
+      - Dry-run:
+          <<: *common_publish_filters
+          requires:
+            - Pre-publish
+      - Confirmation:
+          type: approval
```

Comment:
Question - when a job is put on hold, what happens if we just let that job sit there forever? It doesn’t hold up one of our container slots does it? I’m assuming not, and that we can just leave them be, but I'm wondering if we have to plan for any kind of on-hold cleanup.

Reply:
Second question (more of a wondering out loud comment really) - I wonder if we have unlimited user access for Apollo staff to access Circle's admin, to approve jobs. I know Circle's current plans do, but I think we're on an older grandfathered plan. I'm just envisioning a slight annoyance if people can't access the admin to approve their jobs (kinda like the annoyance we're facing with Netlify). Probably not an issue though.

Reply:
> when a job is put on hold, what happens if we just let that job sit there forever? It doesn’t hold up one of our container slots does it? I’m

Theoretically, the job could hold for ~30 days since [that's how long "workspaces" (which we persist to) are stored](https://circleci.com/docs/2.0/configuration-reference/#persist_to_workspace).  It does _not_ use up a container slot (it's not an actual container and doesn't define an `executor` of any sort; it's a higher level primitive).  After 30 days, the job will be unable to continue.

> unlimited user access for Apollo staff to access Circle's admin, to approve jobs

Any Apollo team member or project collaborator should all have access to "approve" the held job.  There is a new level of granularity that has been recently introduced to CircleCI which we [may want to explore, eventually](https://circleci.com/docs/2.0/contexts/#restrict-a-context-to-a-security-group-or-groups).